src/SignalR/perf/benchmarkapps/BenchmarkServer/Hubs/EchoHub.cs

```diff
@@ -30,6 +35,29 @@ public async Task Broadcast(int duration)
             Console.WriteLine("Broadcast exited: Sent {0} messages", sent);
         }
 
+        public static string Status => $"{_connectionCount} current, {_peakConnectionCount} peak.";
+
+        public void LogConnections(string label) {
+            if (_connectionCount < 1000 || _connectionCount % 100 == 0)
+            {
+                var timeSinceServerStart = DateTime.Now.Subtract(_serverStart).ToString(@"hh\:mm\:ss");
+                Console.WriteLine($"[{timeSinceServerStart}] {label}: {Status}");
+            }
+        }
+
+        public override Task OnConnectedAsync() {
+            _connectionCount++;
+            _peakConnectionCount = Math.Max(_connectionCount, _peakConnectionCount);
```

Comment:
```suggestion
            var savedPeakConnectionCount = Interlocked.Read(ref _peakConnectionCount);

            while (savedConnectionCount > savedPeakConnectionCount) {
                if (Interlocked.CompareExchange(ref _peakConnectionCount, connectionCount, savedPeakConnectionCount) == savedPeakConnectionCount) {
                    break;
                }
                else
                {
                	savedPeakConnectionCount = Interlocked.Read(ref _peakConnectionCount);
                }
            }
```

To do this safely without locking you would need to do something like the above. At this point, it's probably just easier just to track _connectionCount and _peakConnectionCount inside the lock. The lock would be acquired for such short intervals, that I doubt it would cause performance issues.

https://docs.microsoft.com/en-us/dotnet/api/system.threading.interlocked.compareexchange?view=netcore-3.0

Reply:
Paused on this for a while as I needed to read up on `Interlocked`.  Even so, after making all changes (except this one), I agree that using `lock` is probably going to be neater, certainly because in the future I can foresee a `totalDisconnectedCount` or similar.

To my understanding I'll need a single lock object, and I'll use it above and in all other places that I'm currently using `Interlocked` to manipulate those `int` values.

Reply:
> To my understanding I'll need a single lock object, and I'll use it above and in all other places that I'm currently using Interlocked to manipulate those int values.

Just lock all reads and writes to all the static fields you added.

If you want to go for bonus points, you could make a EchoHubConnectionCounter [singleton service](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-2.2) instead of using the statics and register it in the [sample's Startup.ConfigureServices](https://github.com/aspnet/AspNetCore/blob/feff89639c0ddc0b59b90d440db530f936610e8f/src/SignalR/perf/benchmarkapps/BenchmarkServer/Startup.cs#L20). You could then inject the EchoHubConnectionCounter via the EchoHub constructor.

Reply:
switched for `EchoHubConnectionCounter`